### ISSUE TITLE
Misleading error message in tf.broadcast_to

### ISSUE THREAD

**Issue by mjkim720** [tensorflow/tensorflow#41504]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 18.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: N/A
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): v2.2.0-rc4-8-g2b96f3662b 2.2.0
- Python version: 3.7.6
- Bazel version (if compiling from source): N/A
- GCC/Compiler version (if compiling from source): N/A
- CUDA/cuDNN version: N/A
- GPU model and memory: N/A

**Describe the current behavior**
When running `tf.broadcast_to`, the input shown in the error message for an invalid argument is different from the given input.  `[110, 53, 104, 147, 157, 123, 5, 24, 188, 40, 5, 2]` (given) vs `[2,2,2,2,2,2,2,2,2,2,2,2]` (error message).

**Describe the expected behavior**
Correct error message

**Standalone code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.
```python
import tensorflow as tf

x = tf.constant([1, 2, 3])
tf.broadcast_to([x, [110, 53, 104, 147, 157, 123, 5, 24, 188, 40, 5, 2])
```

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.
`InvalidArgumentError: Shape [2,2,2,2,2,2,2,2,2,2,2,2] would have more than 2**63 - 1 elements [Op:BroadcastTo]`

**Comment by lordtt13** [comment]
So the error isn't that misleading, you're getting it because you're broadcasting it to too big a tensor. InvalidArgumentError: Shape [2,2,2,2,2,2,2,2,2,2,2,2] would have more than 2**63 - 1 elements [Op:BroadcastTo] appears when the number of elements has exceeded the given limit. You will have more than 2**63 - 1 elements when you broadcast it to the given shape [110, 53, 104, 147, 157, 123, 5, 24, 188, 40, 5, 2] or the upper limit of shapes [2,2,2,2,2,2,2,2,2,2,2,2].

**Comment by yongtang** [comment]
The issue is that the construction of error message in tensor_shape.cc does not copy all dims of shape, only the current dim of the shape (when error check happens). Created a PR #41514 for the fix.

**Comment by oawxkw** [comment]
I also observed the following API aliases can cause the same issue in older versions of tensorflow.
Users should be cautious when using them on both CPU and GPU up to tensorflow 2.3.4 (v2.3.3-137-gea90cf44f73).

> On GPU, it throws an "CUDA runtime implicit initialization on GPU:0 failed." Exception before tensorflow 2.4.0 (v2.4.0-rc4-71-g582c8d236cb).

- `(tf.broadcast_to)`, `tf.compat.v1.broadcast_to`

<details>
  <summary>Code to reproduce the issue in <code>tf.compat.v1.broadcast_to</code> in older versions</summary>

```python
import tensorflow as tf
print(tf.version.GIT_VERSION, tf.version.VERSION, flush=True)
print(tf.config.list_physical_devices(), flush=True)


try:
    x = tf.constant([1, 2, 3])
    tf.compat.v1.broadcast_to(x, [110, 53, 104, 147, 157, 123, 5, 24, 188, 40, 5, 2])
except Exception as e:
    print("Failed! Error:", str(e), flush=True)
else:
    print("Success!", flush=True)
```

On my CPU machine, it throws an error with a fixed shape <code>[2,2,2,2,2,2,2,2,2,2,2,2]</code> instead of the given shape <code>[110, 53, 104, 147, 157, 123, 5, 24, 188, 40, 5, 2]</code>, which is not expected.

```text
v2.3.3-137-gea90cf44f73 2.3.4
[PhysicalDevice(name='/physical_device:CPU:0', device_type='CPU'), PhysicalDevice(name='/physical_device:XLA_CPU:0', device_type='XLA_CPU'), PhysicalDevice(name='/physical_device:XLA_GPU:0', device_type='XLA_GPU')]
Failed! Error: Shape [2,2,2,2,2,2,2,2,2,2,2,2] would have more than 2**63 - 1 elements [Op:BroadcastTo]
```

While on GPU, it throws a "CUDA runtime implicit initialization on GPU:0 failed." Exception, which is also unexpected.

```text
v2.3.3-137-gea90cf44f73 2.3.4
[PhysicalDevice(name='/physical_device:CPU:0', device_type='CPU'), PhysicalDevice(name='/physical_device:XLA_CPU:0', device_type='XLA_CPU'), PhysicalDevice(name='/physical_device:XLA_GPU:0', device_type='XLA_GPU'), PhysicalDevice(name='/physical_device:GPU:0', device_type='GPU')]
Failed! Error: CUDA runtime implicit initialization on GPU:0 failed. Status: device kernel image is invalid
```
</details>

It seems to be fixed in tensorflow 2.4.0 (v2.4.0-rc4-71-g582c8d236cb) and later versions.
